crates/chain/src/indexed_tx_graph.rs

```diff
@@ -0,0 +1,325 @@
+                    if txout.is_confirmed_and_spendable(tip_height) {
+                        confirmed += txout.txout.value;
+                    } else if !txout.is_mature(tip_height) {
+                        immature += txout.txout.value;
+                    }
+                }
+                ObservedAs::Unconfirmed(_) => {
+                    if should_trust(&txout.txout.script_pubkey) {
+                        trusted_pending += txout.txout.value;
+                    } else {
+                        untrusted_pending += txout.txout.value;
+                    }
+                }
+            }
+        }
+
+        Ok(Balance {
+            immature,
+            trusted_pending,
+            untrusted_pending,
+            confirmed,
+        })
+    }
+
+    pub fn balance<C, F>(&self, chain: &C, chain_tip: BlockId, should_trust: F) -> Balance
+    where
+        C: ChainOracle<Error = Infallible>,
+        F: FnMut(&Script) -> bool,
+    {
+        self.try_balance(chain, chain_tip, should_trust)
+            .expect("error is infallible")
+    }
+}
+
+/// A structure that represents changes to an [`IndexedTxGraph`].
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize, serde::Serialize),
+    serde(
+        crate = "serde_crate",
+        bound(
+            deserialize = "A: Ord + serde::Deserialize<'de>, IA: serde::Deserialize<'de>",
+            serialize = "A: Ord + serde::Serialize, IA: serde::Serialize"
+        )
+    )
+)]
+#[must_use]
+pub struct IndexedAdditions<A, IA> {
+    /// [`TxGraph`] additions.
+    pub graph_additions: Additions<A>,
+    /// [`Indexer`] additions.
+    pub index_additions: IA,
+}
+
+impl<A, IA: Default> Default for IndexedAdditions<A, IA> {
+    fn default() -> Self {
+        Self {
+            graph_additions: Default::default(),
+            index_additions: Default::default(),
+        }
+    }
+}
+
+impl<A: Anchor, IA: Append> Append for IndexedAdditions<A, IA> {
+    fn append(&mut self, other: Self) {
+        self.graph_additions.append(other.graph_additions);
+        self.index_additions.append(other.index_additions);
+    }
+}
+
+/// Represents a structure that can index transaction data.
+pub trait Indexer {
+    /// The resultant "additions" when new transaction data is indexed.
+    type Additions;
+
+    /// Scan and index the given `outpoint` and `txout`.
+    fn index_txout(&mut self, outpoint: OutPoint, txout: &TxOut) -> Self::Additions;
+
+    /// Scan and index the given transaction.
+    fn index_tx(&mut self, tx: &Transaction) -> Self::Additions;
+
+    /// Apply additions to itself.
+    fn apply_additions(&mut self, additions: Self::Additions);
+
+    /// Determines whether the transaction should be included in the index.
+    fn is_tx_relevant(&self, tx: &Transaction) -> bool;
+}
+
+/// A trait that extends [`Indexer`] to also index "owned" script pubkeys.
+pub trait OwnedIndexer: Indexer {
```

Comment:
(nit: this is not a super explicative name)

Reply:
This is a side note, but I've actually started doubting whether we need this trait. It is basically used for filtering only (which can be done in layers above).

Reply:
I'm not sure but I think I mentioned that this trait has the wrong name in my mind. It should at least have `Spk` in the name because it can only answer queries about owning spks. It can just be called `SpkOwner`. There's also no need to require it implements `Indexer`. I would define it as:

```rust
pub trait TxoOwner {
    pub fn owns(&self, tx: &Transaction, vout: u32) -> bool;
}
```

because this is strictly more general and covers silent payments.